### ISSUE TITLE
Expand template contains itself

### ISSUE THREAD

**Issue by xxyzz** [tatuylonen/wikitextprocessor#193]
Page: https://ru.wiktionary.org/wiki/footer
Template: https://ru.wiktionary.org/wiki/Шаблон:длина_слова

The "длина_слова" template calls itself if it is used for substitution, or use "main other" template. I think it's "{{{|safesubst:}}}" in the template delays the expansion of the arguments, but our code doesn't do that and expand the "длина_слова" template recursively.

**Comment by kristian-clausal** [comment]
Oh, that is such a headache... I have no recollection about how we do subst: and safesubst:, but it can't be pretty. The proper way to do it, based on the documentation, would be to actually expand them before anything else... I guess we just do everything in all go.

**Comment by xxyzz** [comment]
Current code ignores them: https://github.com/tatuylonen/wikitextprocessor/blob/958098c50df1a116ee5549f7e4d9352f349265d7/src/wikitextprocessor/core.py#L1403-L1408

Expand before anything else? I think mediawiki does the opposite. And the "длина слова" template expands to a category link, so this bug doesn't affect the ru extractor much.

**Comment by kristian-clausal** [comment]
https://en.wikipedia.org/wiki/Help:Substitution#Technical_implementation
> This means that substitution necessarily occurs before any actions performed at the time of page rendering (conversion of the stored wikitext to HTML). In particular, substitutions are done before transclusions.

Substitutions change the page source itself. Normally templates transclude and are called each time.

**Comment by kristian-clausal** [comment]
I am staring at that template, and my head is starting to hurt. `{{{|safesubst}}}` is a hack used in templates so that you can actually *use* `safesubst:` as a keyword output from a template. If you wrote a template with `{{safesubt:...}}`, it would instantly evaluate when you save the template, removing the `safesubst:` from the source of the template and locking the template into whatever it was evaluated as. So if you want to return `safesubst:` from a template to another template or to the main article, you need to trick the parser into returning `safesubst:` the string by making it the second argument of a triple-braces, and because the first argument is empty / always false the template returns the `safesubst:` as a string inserting it inside the double-braces and outputting `{{safesubst:...}}` on the *next* level...

**Comment by kristian-clausal** [comment]
Ok, there might be a brighter future here:

1. I do not believe that you can find find `{{subst:...}}` on any article page, (unless someone's used that kludge where you have a subst: inside a subst: so that you need to save a page *twice* (or more if it goes even deeper) so that the substitutions take place when the page is saved.
2. safesubst is meant to be usable as both a subst and "not". Basically, safesubst can be *ignored*, which is what we basically do now.

We might be able to get away with small changes... The problem with the template in the first post might not be safesubst, it might be "ifsubst" which checks if we're in "substitution mode" (never applicable for us) or "transclusion mode", and if that breaks that might explain the infinite loop?

**Comment by xxyzz** [comment]
I agree we could continue ignore "safesubst", I guess the problem is the template arguments passed to "#ifeq" are expanded before checking the `if` condition?

Here is a simplified test:

```python
    def test_test(self):
        self.ctx.add_page("Template:ifsubst", 10, """{{#ifeq:yes|yes
 |{{{2|}}}
 |{{{1|}}}
}}""")
        self.ctx.start_page("")
        self.assertEqual(self.ctx.expand("{{ifsubst|yes|no}}"), "no")
        self.assertEqual(self.ctx.expand("{{ifsubst|{{длина слова}}|aaa}}), "aaa")  # fail
```

**Comment by kristian-clausal** [comment]
If we get the behavior of `subst:` right, then `ifsubst` should theoretically work... It relies on the difference between the behavior of `subst` when substituting and transcluding (safesubst has the 'same' behavior for each).

I am still checking all the pages for `{{subst:`, and I've only found one broken one that was left on the page because something weird broke it; couldn't fix it, so removed the malformed template from the article ("leftback" if anyone wants to check it out.) Otherwise, not a single hit, and I've probably grepped through millions of pages by now.

Templates have them, but those templates are supposed to be used once with `{{subst:` (which will enable recursive subst:ing).

**Comment by kristian-clausal** [comment]
I'm an idiot, ifsubst uses safesubst:...
